homeassistant/components/music_assistant/services.py

```diff
@@ -0,0 +1,223 @@
+"""Custom actions (previously known as services) for the Music Assistant integration."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING
+
+from music_assistant_models.enums import MediaType
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntryState
+from homeassistant.core import (
+    HomeAssistant,
+    ServiceCall,
+    ServiceResponse,
+    SupportsResponse,
+    callback,
+)
+from homeassistant.exceptions import HomeAssistantError, ServiceValidationError
+import homeassistant.helpers.config_validation as cv
+
+from .const import (
+    ATTR_ALBUM_ARTISTS_ONLY,
+    ATTR_ALBUM_TYPE,
+    ATTR_ALBUMS,
+    ATTR_ARTISTS,
+    ATTR_CONFIG_ENTRY_ID,
+    ATTR_FAVORITE,
+    ATTR_ITEMS,
+    ATTR_LIBRARY_ONLY,
+    ATTR_LIMIT,
+    ATTR_MEDIA_TYPE,
+    ATTR_OFFSET,
+    ATTR_ORDER_BY,
+    ATTR_PLAYLISTS,
+    ATTR_RADIO,
+    ATTR_SEARCH,
+    ATTR_SEARCH_ALBUM,
+    ATTR_SEARCH_ARTIST,
+    ATTR_SEARCH_NAME,
+    ATTR_TRACKS,
+    DOMAIN,
+)
+from .schemas import (
+    LIBRARY_RESULTS_SCHEMA,
+    SEARCH_RESULT_SCHEMA,
+    media_item_dict_from_mass_item,
+)
+
+if TYPE_CHECKING:
+    from music_assistant_client import MusicAssistantClient
+
+    from . import MusicAssistantConfigEntry
+
+SERVICE_SEARCH = "search"
+SERVICE_GET_LIBRARY = "get_library"
+DEFAULT_OFFSET = 0
+DEFAULT_LIMIT = 25
+DEFAULT_SORT_ORDER = "name"
+
+
+@callback
+def get_music_assistant_client(
+    hass: HomeAssistant, config_entry_id: str
+) -> MusicAssistantClient:
+    """Get the Music Assistant client for the given config entry."""
+    entry: MusicAssistantConfigEntry | None
+    if not (entry := hass.config_entries.async_get_entry(config_entry_id)):
+        raise ServiceValidationError("Entry not found")
+    if entry.state is not ConfigEntryState.LOADED:
+        raise ServiceValidationError("Entry not loaded")
+    return entry.runtime_data.mass
+
+
+@callback
+def register_actions(hass: HomeAssistant) -> None:
+    """Register custom actions."""
+    register_search_action(hass)
+    register_get_library_action(hass)
```

Comment:
I would either have 1 function for just registering actions, not have 2 functions with defining + registering, but this is just a style remark

Reply:
Let's assume that maybe more services will be added in the future. That will get us a super lengthy function. I like the split up approach better style-wise.

Reply:
FYI, you can now also make the service calls a standalone function, since `hass` is now included in the service call object. This way you can just have all the registrations in one place

Reply:
Nice!